### PR TITLE
New Post - XML-NewEmployees.md

### PR DESCRIPTION
**Pull request by samryanturner**
Post on how to leverage xml with new user scripts in PowerShell. Questions to @samryanturner

# PR Summary

## PR Checklist

- Do you want this post to be published on a specific date?: Y/N [Specify date]:
- [x] I have read the [contributors guide][contrib] and followed the style and process guidelines
- [x] PR has a meaningful title
- [ ] Includes content related an open issue
  - see [Closing issues using keywords][key].
- [x] This PR is ready to merge and is not **Work in Progress**

[contrib]: https://docs.microsoft.com/powershell/scripting/community/contributing/overview
[key]: https://help.github.com/en/articles/closing-issues-using-keywords

### CODE REVIEW

**Comment by samryanturner** [comment]
Let me know what changes are needed 😄 . I was thinking of adding a last section showing the value itself with the content from the xml.

I also noticed that the actual dot-notation part doesn't actually reference the correct elements in the xml made for the demo. I can fix that.

**Comment by doctordns** [comment]
A well written article. The examples look good too!

But I kept asking what is the business problem being solved here? I'd like to see more about why you use XML and what value this approach has over other text-based methods. Using XML writers (or any .NET class for that matter) is only done when there are no PowerShell commands. A paragraph at the start of this explaining why XML would be good

File: Posts/2021/02/XML-NewEmployees.md [hunk 1]
```diff
@@ -0,0 +1,180 @@
...
+				<Postcode>B1 1BB</Postcode>
+				<Country>GB</Country>
+			</Birmingham>
+		</UK-Offices>
+	</Company-1>
+</Companies>
+```
+## PowerShell - Working With XML
+
+So, now we have some useable data to work with PowerShell on. You can add more elements for Offices/Companies you are creating users for. It will become clear how you reference this next.
```
Comment (doctordns) [inline Posts/2021/02/XML-NewEmployees.md:92]: two tings>
1. in MD seperate each sentence on a separate line. THat makes it really a lot easier to edit./
2.  Make second sentence read. "It becomes clear how you reference this next." avoid the use of future tense.

File: Posts/2021/02/XML-NewEmployees.md [hunk 2]
```diff
@@ -0,0 +1,180 @@
...
+Dead simple. 4 questions with 4 answers, into 4 variables. Instead of asking lots of questions to find out information like postcode or country, let's use our XML!
+
+```powershell
+# Declare some more variables
+$FirstInitial = "$FirstName.substring(0,1)"
+$Path = 'OU=Users,OU=Contoso,DC=Contoso,DC=com'
+
+# Build hashtable of new user parameters
+  $NewUserParams =@{
+
```
Comment (doctordns) [inline Posts/2021/02/XML-NewEmployees.md:133]: remove blank line

**Comment by samryanturner** [comment]
Thanks Thomas! Sure I can add something in.

Generally I think having a central source of address information for offices etc. is much more efficient and cleaner than just storing plain text strings in scripts.

It also allows other people in the business who are not so comfortable editing PowerShell scripts to amend this information.

File: Posts/2021/02/XML-NewEmployees.md [hunk 3]
```diff
@@ -0,0 +1,180 @@
...
+```powershell
+
+        StreetAddress = ($XML.Companies.$Company.$office.Street).Trim()
+        PostalCode = ($XML.Companies.$Company.$Office.Postcode).Trim()
+        City = ($XML.Companies.$Company.$Office.City).Trim()
+        State = ($xml.Companies.$Company.$Office.State).Trim()
+        Country = ($xml.Companies.$Company.$Office.Country).Trim()
+```
+# Helpful Notes
+* Be careful when creating your elements! No whitespace allowed!
```
Comment (doctordns) [inline Posts/2021/02/XML-NewEmployees.md:166]: can you explain a bit more?  I think you mean XML Elements. WHERE are whitespace not allowed?

File: Posts/2021/02/XML-NewEmployees.md [hunk 4]
```diff
@@ -0,0 +1,180 @@
...
+You can really flesh this out more and get a load of information into XML for referencing, e.g. company project information, sets of AD groups users could be added to, managed bookmark URLS (to then be copied into registry items). 
+
+The list goes on and on: they are just strings at the end of the day! 
+
+I was inspired by the following people to both do this work and write this blog post.
+
+* Thomas Lee (https://twitter.com/doctordns) - thanks for getting this blog started in the first place!
+* Adam Bertram (https://twitter.com/adbertram) and his fantastic book "PowerShell for SysAdmins: Workflow Automation Made Easy" (https://nostarch.com/powershellsysadmins)
+* Tobias Weltner (*I couldn't find your twitter :(*) and his article on Mastering everday XML Tasks in PowerShell (https://www.powershellmagazine.com/2013/08/19/mastering-everyday-xml-tasks-in-powershell/)
+
```
Comment (doctordns) [inline Posts/2021/02/XML-NewEmployees.md:180]: restyle the URLs as noted above.
Also, Tobias's Twitter handle is **@TobiasPSP** (and no relation to my company PS Partnership !).

File: Posts/2021/02/XML-NewEmployees.md [hunk 5]
```diff
@@ -0,0 +1,180 @@
...
+You have multiple subsidiary companies that each have their own individual sets of offices and addresses. You need to ensure that each new user you create has the correct address and contact information based on where they are working.
+
+## Getting Started - Creating the XML
+
+To create the xml, we're going to use the XmlWriter class, put this into a variable to build upon and add some formatting options. Like so:
+
+```powershell
+$XMLPath = "C:\Users\sam\AD-References.xml"
+$XMLWriter = New-Object System.XML.XMlTextWriter($XMLPath,$null)
```
Comment (doctordns) [inline Posts/2021/02/XML-NewEmployees.md:26]: perhaps add a reference to what an XMLTextWriter is and what you are going to do with it. Also possibly add a URL reference to  [XMLTextWriterClass](https://docs.microsoft.com/en-us/dotnet/api/system.xml.xmltextwriter?view=net-5.0)

File: Posts/2021/02/XML-NewEmployees.md [hunk 6]
```diff
@@ -0,0 +1,180 @@
+```yaml
+---
+post_title: 'Leveraging XML with New Employee PowerShell Scripts'
+username: s-r-turner
+categories: ActiveDirectory, XML, PowerShell
+tags: XML, ActiveDirectory
+featured_image:
+summary: using xml referencing to bolster up your new employee scripts with information.
```
Comment (doctordns) [inline Posts/2021/02/XML-NewEmployees.md:8]: summary: Using XML referencing to bolster up your new employee scripts with information.

File: Posts/2021/02/XML-NewEmployees.md [hunk 7]
```diff
@@ -0,0 +1,180 @@
+```yaml
+---
+post_title: 'Leveraging XML with New Employee PowerShell Scripts'
+username: s-r-turner
+categories: ActiveDirectory, XML, PowerShell
```
Comment (doctordns) [inline Posts/2021/02/XML-NewEmployees.md:5]: categories: PowerShell

File: Posts/2021/02/XML-NewEmployees.md [hunk 8]
```diff
@@ -0,0 +1,180 @@
...
+tags: XML, ActiveDirectory
+featured_image:
+summary: using xml referencing to bolster up your new employee scripts with information.
+---
+```
+
+# Leveraging XML With New Employee PowerShell Scripts
+
+I'm going to show how you can leverage XML files for referencing information, to help bolster up your Active Directory user accounts.
```
Comment (doctordns) [inline Posts/2021/02/XML-NewEmployees.md:14]: What information and why do you need to olster AD User Accounts. 
Ensure you make the business case for why you are recommending this approach.

File: Posts/2021/02/XML-NewEmployees.md [hunk 9]
```diff
@@ -0,0 +1,180 @@
...
+
+## Scenario
+
+You have multiple subsidiary companies that each have their own individual sets of offices and addresses. You need to ensure that each new user you create has the correct address and contact information based on where they are working.
```
Comment (doctordns) [inline Posts/2021/02/XML-NewEmployees.md:18]: Givn this, I assume what you are going to is to create some XML to contain user addresses and later use that to update AD User Objects. 

Is it worth mentioning you could so something similar to AD Computers too?

File: Posts/2021/02/XML-NewEmployees.md [hunk 10]
```diff
@@ -0,0 +1,180 @@
...
+## Getting Started - Creating the XML
+
+To create the xml, we're going to use the XmlWriter class, put this into a variable to build upon and add some formatting options. Like so:
+
+```powershell
+$XMLPath = "C:\Users\sam\AD-References.xml"
+$XMLWriter = New-Object System.XML.XMlTextWriter($XMLPath,$null)
+$xmlWriter.Formatting = 'Indented'
+$xmlWriter.Indentation = 1
+$XmlWriter.IndentChar = "`t"
```
Comment (doctordns) [inline Posts/2021/02/XML-NewEmployees.md:29]: is it worth explaining why you use these?

File: Posts/2021/02/XML-NewEmployees.md [hunk 11]
```diff
@@ -0,0 +1,180 @@
...
+			<Birmingham>
+				<Street>Unit 77, 132 Dummy Lane</Street>
+				<City>Birmingham</City>
+				<State>Staffordshire</State>
+				<Postcode>B1 1BB</Postcode>
+				<Country>GB</Country>
+			</Birmingham>
+		</UK-Offices>
+	</Company-1>
+</Companies>
```
Comment (doctordns) [inline Posts/2021/02/XML-NewEmployees.md:88]: I thought indentation was 1 character, this looks like more?
can you shrink down the indentation a bit - say to 2 characters

File: Posts/2021/02/XML-NewEmployees.md [hunk 12]
```diff
@@ -0,0 +1,180 @@
...
+    
+    if (Test-Path $XMLPath) {
+        
+        $XML = New-Object -TypeName XML
+        $XML.Load($XMLPath)
+    }
+    else {
+        Start-Process "$PSScriptRoot\Write-XML.ps1" -Wait
+        $XML = New-Object -TypeName XML
+        $XML.Load($XMLPath)
```
Comment (doctordns) [inline Posts/2021/02/XML-NewEmployees.md:110]: # A simpler way?
$XML  = [XML] (Get-Content $XMLPath)

File: Posts/2021/02/XML-NewEmployees.md [hunk 13]
```diff
@@ -0,0 +1,180 @@
...
+        Enabled = $true
+    }
+    
+# Create the user
+New-ADUser @NewUserParams
+```
+Simple as that, browsing through the xml elements using dot-notation, and trimming off any unwanted whitespace. This is cropped below for convenience:
+
+```powershell
+
```
Comment (doctordns) [inline Posts/2021/02/XML-NewEmployees.md:158]: remove blank line

File: Posts/2021/02/XML-NewEmployees.md [hunk 14]
```diff
@@ -0,0 +1,180 @@
...
+        StreetAddress = ($XML.Companies.$Company.$office.Street).Trim()
+        PostalCode = ($XML.Companies.$Company.$Office.Postcode).Trim()
+        City = ($XML.Companies.$Company.$Office.City).Trim()
+        State = ($xml.Companies.$Company.$Office.State).Trim()
+        Country = ($xml.Companies.$Company.$Office.Country).Trim()
+```
+# Helpful Notes
+* Be careful when creating your elements! No whitespace allowed!
+* Your inputted variables using Read-Host need to match the references in the XML. You could make use of Validation Scripts to validate your inputs. Here is a helpful read on everything to do with validating inputs: https://adamtheautomator.com/powershell-validatescript/
```
Comment (doctordns) [inline Posts/2021/02/XML-NewEmployees.md:167]: change the start of the sentence to "The variables you input using `Read-Host` need to match"

Also, fix URL  "... Here is a helpful read on [everything to do with validation inputs.][https://adamtheautomator.com/powershell-validatescript/]"

**Review by doctordns — COMMENTED** [review]
Some minor points in this good article.

File: Posts/2021/02/XML-NewEmployees.md (hunk 12 above)
Comment (samryanturner) [reply Posts/2021/02/XML-NewEmployees.md:110]: instead of line 110? I like it! 👍

File: Posts/2021/02/XML-NewEmployees.md (hunk 11 above)
Comment (samryanturner) [reply Posts/2021/02/XML-NewEmployees.md:88]: 2 will indent further. I think the issue is the number of root elements in the XML. I can cut it down for the purpose of the post.

**Comment by samryanturner** [comment]
@doctordns changes completed, let me know your thoughts! :)

File: Posts/2021/02/XML-NewEmployees.md [hunk 15]
```diff
@@ -0,0 +1,175 @@
+```yaml
+---
+post_title: 'Leveraging XML with New Employee PowerShell Scripts'
+username: s-r-turner
+categories: PowerShell
+tags: XML, ActiveDirectory
+featured_image:
+summary: using XML referencing to bolster up your new employee scripts with information.
+---
+```
```
Comment (sdwheeler) [inline Posts/2021/02/XML-NewEmployees.md:10]: The YAML is part of the post, not a code block..
```suggestion
---
post_title: 'Leveraging XML with New Employee PowerShell Scripts'
username: s-r-turner
categories: PowerShell
tags: XML, ActiveDirectory
featured_image:
summary: using XML referencing to bolster up your new employee scripts with information.
---
```

File: Posts/2021/02/XML-NewEmployees.md [hunk 16]
```diff
@@ -0,0 +1,175 @@
...
+## Scenario
+
+You have multiple subsidiary companies that each have their own individual sets of offices and addresses. 
+You need to ensure that each new user you create has the correct address and contact information based on where they are working.
+Whilst in this post I am talking about User Accounts, there is no reason why you can't apply the same logic for Computer Accounts or Resources. 
+No one likes asking around for the location of a meeting room you were supposed to be in 10 minutes ago!
+
+## Getting Started - Creating the XML
+
+To create the xml, we're going to use the [XmlWriter](https://docs.microsoft.com/en-us/dotnet/api/system.xml.xmlwriter?view=net-5.0) class, put this into a variable to build upon and add some formatting options (this makes it easier to see groups of information). Like so:
```
Comment (sdwheeler) [inline Posts/2021/02/XML-NewEmployees.md:27]: Always remove the `?view=<version>` and the locale `en-us/` from URL on docs.microsoft.com.

```suggestion
To create the xml, we're going to use the [XmlWriter](https://docs.microsoft.com/dotnet/api/system.xml.xmlwriter) class, put this into a variable to build upon and add some formatting options (this makes it easier to see groups of information). Like so:
```

File: Posts/2021/02/XML-NewEmployees.md [hunk 17]
```diff
@@ -0,0 +1,175 @@
...
+username: s-r-turner
+categories: PowerShell
+tags: XML, ActiveDirectory
+featured_image:
+summary: using XML referencing to bolster up your new employee scripts with information.
+---
+```
+
+# Leveraging XML With New Employee PowerShell Scripts
+
```
Comment (sdwheeler) [inline Posts/2021/02/XML-NewEmployees.md:13]: Remove the H1 title here. There should be no H1 headers. The title is automatically added to the post from the `post_title`  metadata.

File: Posts/2021/02/XML-NewEmployees.md [hunk 18]
```diff
@@ -0,0 +1,175 @@
...
+To create the xml, we're going to use the [XmlWriter](https://docs.microsoft.com/en-us/dotnet/api/system.xml.xmlwriter?view=net-5.0) class, put this into a variable to build upon and add some formatting options (this makes it easier to see groups of information). Like so:
+
+```powershell
+$XMLPath = "C:\Users\sam\AD-References.xml"
+$XMLWriter = New-Object System.XML.XMlTextWriter($XMLPath,$null)
+$xmlWriter.Formatting = 'Indented'
+$xmlWriter.Indentation = 1
+$XmlWriter.IndentChar = "`t"
+```
+If you run that, nothing crazy happens. We have a new xml source file, with nothing inside; we have set ourselves up to flesh it out.
```
Comment (sdwheeler) [inline Posts/2021/02/XML-NewEmployees.md:36]: Blank lines before and after all markdown blocks (code, paragraphs, lists, tables, etc.)

```suggestion

If you run that, nothing crazy happens. We have a new xml source file, with nothing inside; we have set ourselves up to flesh it out.
```

File: Posts/2021/02/XML-NewEmployees.md [hunk 19]
```diff
@@ -0,0 +1,175 @@
...
+
+```powershell
+# write the header
+$xmlWriter.WriteStartDocument()
+# set XSL statements
+$xmlWriter.WriteProcessingInstruction("xml-stylesheet", "type='text/xsl' href='style.xsl'")
+$XMLWriter.WriteStartElement('Companies')
+```
+Okay awesome, now we're getting somewhere. We have started our "Companies" element. Let's create a company and add some office information beneath that. Then we'll finalise the XML document.
```
Comment (sdwheeler) [inline Posts/2021/02/XML-NewEmployees.md:45]: Insert blank line

```suggestion

Okay awesome, now we're getting somewhere. We have started our "Companies" element. Let's create a company and add some office information beneath that. Then we'll finalise the XML document.
```

File: Posts/2021/02/XML-NewEmployees.md [hunk 20]
```diff
@@ -0,0 +1,175 @@
...
+			<Street>Unit 77, 132 Dummy Lane</Street>
+			<City>Birmingham</City>
+			<State>Staffordshire</State>
+			<Postcode>B1 1BB</Postcode>
+			<Country>GB</Country>
+		</Birmingham>
+	</Company-1>
+</Companies>
+```
+## PowerShell - Working With XML
```
Comment (sdwheeler) [inline Posts/2021/02/XML-NewEmployees.md:87]: Insert blank line

```suggestion

## PowerShell - Working With XML
```

File: Posts/2021/02/XML-NewEmployees.md [hunk 21]
```diff
@@ -0,0 +1,175 @@
...
+        $XML = [XML] (Get-Content $XMLPath)
+    }
+    else {
+        Start-Process "$PSScriptRoot\Write-XML.ps1" -Wait
+        $XML = New-Object -TypeName XML
+        $XML = [XML] (Get-Content $XMLPath)
+    }
+    
+```
+Here, we are testing the path for the XML and, if it doesn't exist, running the creation script again. Then, we're loading it into a variable for reference.
```
Comment (sdwheeler) [inline Posts/2021/02/XML-NewEmployees.md:111]: Insert blank line

```suggestion

Here, we are testing the path for the XML and, if it doesn't exist, running the creation script again. Then, we're loading it into a variable for reference.
```

File: Posts/2021/02/XML-NewEmployees.md [hunk 22]
```diff
@@ -0,0 +1,175 @@
...
+        State = ($XML.Companies.$Company.$Office.State).Trim()
+        Country = ($XML.Companies.$Company.$Office.Country).Trim()
+        Path = $Path
+        ChangePasswordAtLogon = $false
+        Enabled = $true
+    }
+# Create the user
+New-ADUser @NewUserParams
+```
+Simple as that, browsing through the xml elements using dot-notation, and trimming off any unwanted whitespace. This is cropped below for convenience:
```
Comment (sdwheeler) [inline Posts/2021/02/XML-NewEmployees.md:151]: Insert blank lines

```suggestion

Simple as that, browsing through the xml elements using dot-notation, and trimming off any unwanted whitespace. This is cropped below for convenience:

```

File: Posts/2021/02/XML-NewEmployees.md [hunk 23]
```diff
@@ -0,0 +1,175 @@
...
+```powershell
+StreetAddress = ($XML.Companies.$Company.$Office.Street).Trim()
+PostalCode = ($XML.Companies.$Company.$Office.Postcode).Trim()
+City = ($XML.Companies.$Company.$Office.City).Trim()
+State = ($xml.Companies.$Company.$Office.State).Trim()
+Country = ($xml.Companies.$Company.$Office.Country).Trim()
+```
+# Helpful Notes
+* Be careful when creating your XML elements! No whitespace allowed! You will get XML Writer errors if there are spaces in your XML Elements. I use "-" to seperate mine.
```
Comment (sdwheeler) [inline Posts/2021/02/XML-NewEmployees.md:160]: Insert blank lines and change to H2

```suggestion

# Helpful Notes

* Be careful when creating your XML elements! No whitespace allowed! You will get XML Writer errors if there are spaces in your XML Elements. I use "-" to seperate mine.

```

File: Posts/2021/02/XML-NewEmployees.md [hunk 24]
```diff
@@ -0,0 +1,175 @@
...
+```powershell
+$XMLWriter.WriteStartElement('Company-1')
+```
+* The variables you input using `Read-Host` need to match the Element names in the XML. You could make use of Validation Scripts to validate your inputs. Here is a helpful read on [everything to do with validation inputs.](https://adamtheautomator.com/powershell-validatescript)
+# Final Comments and Credits
```
Comment (sdwheeler) [inline Posts/2021/02/XML-NewEmployees.md:165]: Insert blank lines and change to H2

```suggestion

* The variables you input using `Read-Host` need to match the Element names in the XML. You could make use of Validation Scripts to validate your inputs. Here is a helpful read on [everything to do with validation inputs.](https://adamtheautomator.com/powershell-validatescript)

## Final Comments and Credits
```

**Review by sdwheeler — CHANGES_REQUESTED** [review]

**Review by sdwheeler — APPROVED** [review]

**Comment by samryanturner** [comment]
Noticed all the "/" have been removed from all the code blocks. Do we know how to resolve this? @sdwheeler

**Comment by sdwheeler** [comment]
@samryanturner Fixed. There is something wrong with how the markdown get converted for WordPress. I am filing a bug on that.

**Comment by samryanturner** [comment]
That's cool! I figured it was something Wordpress-y. Thank you so much for fixing it!

**Pull request merged**